Sources/CombineLongPolling/LongPollingPublisher.swift

```diff
@@ -1,94 +1,115 @@
-    /// `LongPollingPublisher` will `Never` fail.
-    public struct LongPollingPublisher: Publisher {
-        /// The kind of values published by this publisher.
-        public typealias Output = Result<(Data, URLResponse), URLError>
-
-        /// The kind of errors this publisher might publish.
-        ///
-        /// Use `Never` if this `Publisher` does not publish errors.
-        public typealias Failure = Never
-
-        public let upstream: URLSession.DataTaskPublisher
-
-        /// Creates a new `LongPollingPublisher` from an upstream `DataTaskPublisher`
-        /// - Parameter upstream: the `DataTaskPublisher` containing the `URLRequest`
-        public init(upstream: URLSession.DataTaskPublisher) {
-            self.upstream = upstream
+        func request(_ demand: Subscribers.Demand) {
+            guard let buffer = self.buffer else { return }
+
+            lock.lock()
+
+            if !started && demand > .none {
+                // There's demand, and it's the first demanded value, so we start polling
+                started = true
+                lock.unlock()
+
+                start()
+            } else {
+                lock.unlock()
+            }
+
+            // Flush buffer
+            // If subscriber asked for 10 but we had only 3 in the buffer, it will return 7 representing the remaining demand
+            // We actually don't care about that number, as once we buffer more items they will be flushed right away, so simply ignore it
+            _ = buffer.demand(demand)
         }
 
-        /// This function is called to attach the specified `Subscriber` to this `Publisher` by `subscribe(_:)`
-        ///
-        /// - SeeAlso: `subscribe(_:)`
-        /// - Parameters:
-        ///     - subscriber: The subscriber to attach to this `Publisher`.
-        ///                   once attached it can begin to receive values.
-        public func receive<S>(subscriber: S) where S : Subscriber, S.Failure == Failure, S.Input == Output {
-            start().subscribe(subscriber)
+        public func cancel() {
+            buffer?.complete(completion: .finished)
+            started = false
+            currentRequest = nil
+            buffer = nil
+            self.semaphore = DispatchSemaphore(value: 1)
         }
 
-        private func start() -> AnyPublisher<Output, Never> {
-            upstream
-                .map(Result<(Data, URLResponse), URLError>.success)
-                .catch { error in
-                    Just(.failure(error))
-                }
-                .assertNoFailure()
-                .flatMap(maxPublishers: .max(1)) { event in
-                    self.start().prepend(event)
+        private func start() {
+            self.semaphore = DispatchSemaphore(value: 1)
+            startPolling()
+        }
+
+        private func startPolling() {
+            DispatchQueue.global(qos: .utility).async {
```

Comment:
I think you should create your own queue here. As far I've understood the docs last week, this will return `a global queue`, which other components could try to use. And you are blocking it here.

Another case: Multiple long polling calls at the same time: Do they try to use the same queue, blocking each other?

Reply:
It doesn't block because it's a concurrent queue. Each of them will have a thread waiting for the response.
We can create our own, as long as it's concurrent as well and use low QoS to avoid giving too much priority for these threads, that will be most of the time waiting in the semaphore.

Reply:
I see. Fine then :)